File: src/minigames/treasure_hunter/java/work/lclpnet/ap2/game/treasure_hunter/TreasureHunterInstance.java
package work.lclpnet.ap2.game.treasure_hunter;

import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.item.ItemEntity;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.item.enchantment.Enchantments;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.gamerules.GameRules;
import work.lclpnet.ap2.api.base.Participants;
import work.lclpnet.ap2.api.game.MiniGameHandle;
import work.lclpnet.ap2.api.game.data.DataContainer;
import work.lclpnet.ap2.impl.game.FFAGameInstance;
import work.lclpnet.ap2.impl.game.data.CombinedDataContainer;
import work.lclpnet.ap2.impl.game.data.IntScoreDataContainer;
import work.lclpnet.ap2.impl.game.data.OrderedDataContainer;
import work.lclpnet.ap2.impl.game.data.type.PlayerRef;
import work.lclpnet.ap2.impl.map.MapUtil;
import work.lclpnet.ap2.impl.util.ItemHelper;
import work.lclpnet.gaco.ds.BlockBox;
import work.lclpnet.kibu.access.entity.PlayerInventoryAccess;
import work.lclpnet.kibu.access.entity.ServerPlayerAccess;
import work.lclpnet.kibu.hook.HookRegistrar;
import work.lclpnet.kibu.hook.entity.PlayerInteractionHooks;
import work.lclpnet.kibu.translate.Translations;
import work.lclpnet.lobby.game.impl.prot.ProtectionTypes;

import java.util.*;

import static work.lclpnet.ap2.impl.util.ItemHelper.unbreakable;

public class TreasureHunterInstance extends FFAGameInstance {

    private static final float COIN_CHANCE = 0.025f;
    private final Random random = new Random();
    private final OrderedDataContainer<ServerPlayer, PlayerRef> foundChest = new OrderedDataContainer<>(PlayerRef::create);
    private final IntScoreDataContainer<ServerPlayer, PlayerRef> score = new IntScoreDataContainer<>(PlayerRef::create);
    private final CombinedDataContainer<ServerPlayer, PlayerRef> data = new CombinedDataContainer<>(List.of(foundChest, score));
    private final Set<BlockState> materials = new HashSet<>();

    public TreasureHunterInstance(MiniGameHandle gameHandle) {
        super(gameHandle);

        useSurvivalMode();
    }

    @Override
    protected DataContainer<ServerPlayer, PlayerRef> getData() {
        return data;
    }

    @Override
    protected void prepare() {
        commons().gameRuleBuilder()
                .set(GameRules.BLOCK_DROPS, false)
                .set(GameRules.ENTITY_DROPS, false);

        MapUtil.readBlockStates(getMap().requireProperty("materials"), materials, gameHandle.getLogger());

        Participants participants = gameHandle.getParticipants();
        Translations translations = gameHandle.getTranslations();
        HookRegistrar hooks = gameHandle.getHooks();

        hooks.registerHook(PlayerInteractionHooks.USE_BLOCK, (player, world, hand, hitResult) -> {
            if (!(player instanceof ServerPlayer serverPlayer) || !participants.isParticipating(serverPlayer)
                || !world.getBlockState(hitResult.getBlockPos()).is(Blocks.CHEST)) {
                return InteractionResult.PASS;
            }

            if (winManager.isGameOver()) {
                return InteractionResult.FAIL;
            }

            ServerPlayerAccess.playSoundToPlayer(serverPlayer, SoundEvents.AMETHYST_BLOCK_RESONATE, SoundSource.BLOCKS, 1.2f, 1.8f);
            ServerPlayerAccess.playSoundToPlayer(serverPlayer, SoundEvents.CHEST_LOCKED, SoundSource.BLOCKS, 0.2f, 0.5f);

            var scoreEntry = score.getEntry(serverPlayer)
                    .<Object>map(entry -> entry.toText(translations))
                    .orElse(Component.literal("-"));

            var detail = translations.translateText("game.ap2.treasure_hunter.found_treasure", scoreEntry);
            foundChest.add(serverPlayer, detail);
            winManager.complete();

            return InteractionResult.SUCCESS_SERVER;
        });

        hooks.registerHook(PlayerInteractionHooks.BREAK_BLOCK, (world, player, pos, state, blockEntity) -> {
            if (!(player instanceof ServerPlayer serverPlayer) || !participants.isParticipating(serverPlayer)) {
                return true;
            }

            if (!materials.contains(state) || !(random.nextFloat() < COIN_CHANCE)) return true;

            spawnCoin(pos, world);

            return true;
        });

        placeChest();

        useTaskDisplay();
    }

    @Override
    protected void go() {
        gameHandle.protect(config -> {
            config.allow(ProtectionTypes.BREAK_BLOCKS, (entity, pos) -> {
                Level world = entity.level();
                BlockState state = world.getBlockState(pos);

                return materials.contains(state);
            });

            config.allow(ProtectionTypes.PICKUP_ITEM, (player, item) -> {
                if (player instanceof ServerPlayer serverPlayer && item.getItem().is(Items.SUNFLOWER)) {
                    item.discard();
                    giveCoin(serverPlayer);
                }

                return false;
            });
        });

        giveShovelsToPlayers();
    }

    private void spawnCoin(BlockPos pos, Level world) {
        ItemEntity coin = new ItemEntity(world, pos.getX(), pos.getY(), pos.getZ(), new ItemStack(Items.SUNFLOWER));
        world.addFreshEntity(coin);
    }

    private void giveCoin(ServerPlayer player) {
        ServerPlayerAccess.playSoundToPlayer(player, SoundEvents.ARROW_HIT_PLAYER, SoundSource.BLOCKS, 0.7f, 1.55f);

        commons().addScore(player, 1, score);
    }

    private void giveShovelsToPlayers() {
        var efficiency = ItemHelper.getEnchantment(Enchantments.EFFICIENCY, getWorld().registryAccess());

        for (ServerPlayer player : gameHandle.getParticipants()) {
            ItemStack stack = unbreakable(new ItemStack(Items.IRON_SHOVEL));

            stack.enchant(efficiency, 4);

            Inventory inventory = player.getInventory();
            inventory.setItem(4, stack);
            PlayerInventoryAccess.setSelectedSlot(player, 4);
        }
    }

    private void placeChest() {
        ServerLevel world = getWorld();
        BlockBox box = MapUtil.readBox(getMap().requireProperty("chest-area"));
        List<BlockPos> chestAreaList = new ArrayList<>();

        for (BlockPos block : box) {
            BlockState state = world.getBlockState(block);

            if (materials.contains(state)) {
                chestAreaList.add(block.immutable());
            }
        }

        int randomIndex = random.nextInt(chestAreaList.size());

        BlockPos chestPos = chestAreaList.get(randomIndex);
        world.setBlockAndUpdate(chestPos, Blocks.CHEST.defaultBlockState());
    }
}

File: src/minigames/treasure_hunter/src/main/kotlin/work/lclpnet/ap2/game/treasure_hunter/TreasureHunterInstance.kt
package work.lclpnet.ap2.game.treasure_hunter

import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.sounds.SoundEvents
import net.minecraft.sounds.SoundSource
import net.minecraft.world.InteractionResult
import net.minecraft.world.entity.item.ItemEntity
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.Items
import net.minecraft.world.item.enchantment.Enchantments
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.gamerules.GameRules
import work.lclpnet.ap2.api.stats.CommonStats.BlocksBroken
import work.lclpnet.ap2.ext.hooks
import work.lclpnet.ap2.ext.mc.isOf
import work.lclpnet.ap2.game.MiniGameHandle
import work.lclpnet.ap2.game.base.FFAGameInstance
import work.lclpnet.ap2.game.data.CombinedDataContainer
import work.lclpnet.ap2.game.data.IntScoreDataContainer
import work.lclpnet.ap2.game.data.OrderedDataContainer
import work.lclpnet.ap2.game.data.type.PlayerRef
import work.lclpnet.ap2.game.util.createStats
import work.lclpnet.ap2.game.util.useSurvivalMode
import work.lclpnet.ap2.game.util.useTaskDisplay
import work.lclpnet.ap2.impl.map.MapUtil
import work.lclpnet.ap2.impl.util.ItemHelper
import work.lclpnet.ap2.impl.util.ItemHelper.unbreakable
import work.lclpnet.game.impl.prot.ProtectionTypes
import work.lclpnet.game.map.GameMap
import work.lclpnet.kibu.access.entity.PlayerInventoryAccess
import work.lclpnet.kibu.access.entity.ServerPlayerAccess
import work.lclpnet.kibu.hook.entity.PlayerInteractionHooks
import java.util.*

private const val COIN_CHANCE = 0.025f


class TreasureHunterInstance(gameHandle: MiniGameHandle, level: ServerLevel, map: GameMap) : FFAGameInstance(gameHandle, level, map) {

    private val random = Random()
    private val foundChest = OrderedDataContainer(PlayerRef::create)
    private val score = IntScoreDataContainer(PlayerRef::create)
    override val data = CombinedDataContainer(listOf(foundChest, score))
    private val materials = HashSet<BlockState>()
    private val stats = createStats(winManager, score, BlocksBroken)

    init {
        useSurvivalMode()
    }

    override fun prepare() {
        commons().gameRuleBuilder()
            .set(GameRules.BLOCK_DROPS, false)
            .set(GameRules.ENTITY_DROPS, false)

        MapUtil.readBlockStates(map.requireProperty("materials"), materials, gameHandle.logger)

        val participants = gameHandle.participants
        val translations = gameHandle.translations

        PlayerInteractionHooks.USE_BLOCK.registerWith(hooks) { player, world, _, hitResult ->
            if (player !is ServerPlayer || !participants.isParticipating(player)
                    || !world.getBlockState(hitResult.blockPos).isOf(Blocks.CHEST)) {
                return@registerWith InteractionResult.PASS
            }

            if (winManager.gameOver) {
                return@registerWith InteractionResult.FAIL
            }

            ServerPlayerAccess.playSoundToPlayer(player, SoundEvents.AMETHYST_BLOCK_RESONATE, SoundSource.BLOCKS, 1.2f, 1.8f)
            ServerPlayerAccess.playSoundToPlayer(player, SoundEvents.CHEST_LOCKED, SoundSource.BLOCKS, 0.2f, 0.5f)

            val scoreEntry = score.getEntry(player)
                ?.toText(translations)
                ?: Component.literal("-")

            val detail = translations.translateText("game.ap2.treasure_hunter.found_treasure", scoreEntry)
            foundChest.add(player, detail)
            winManager.complete()

            InteractionResult.SUCCESS_SERVER
        }

        PlayerInteractionHooks.BREAK_BLOCK.registerWith(hooks) { world, player, pos, state, _ ->
            if (player !is ServerPlayer || !participants.isParticipating(player)) {
                return@registerWith true
            }

            stats.increment(player, BlocksBroken)

            if (materials.contains(state) && random.nextFloat() < COIN_CHANCE) {
                spawnCoin(pos, world)
            }

            true
        }

        placeChest()

        useTaskDisplay()
    }

    override fun go() {
        gameHandle.protect { config ->
            ProtectionTypes.BREAK_BLOCKS.allow(config) { entity, pos ->
                val state = entity.level().getBlockState(pos)
                materials.contains(state)
            }

            ProtectionTypes.PICKUP_ITEM.allow(config) { player, item ->
                if (player is ServerPlayer && item.item.isOf(Items.SUNFLOWER)) {
                    item.discard()
                    giveCoin(player)
                }

                false
            }
        }

        giveShovelsToPlayers()
    }

    private fun spawnCoin(pos: BlockPos, world: Level) {
        val coin = ItemEntity(world, pos.x.toDouble(), pos.y.toDouble(), pos.z.toDouble(), ItemStack(Items.SUNFLOWER))
        world.addFreshEntity(coin)
    }

    private fun giveCoin(player: ServerPlayer) {
        ServerPlayerAccess.playSoundToPlayer(player, SoundEvents.ARROW_HIT_PLAYER, SoundSource.BLOCKS, 0.7f, 1.55f)
        commons().addScore(player, 1, score)
    }

    private fun giveShovelsToPlayers() {
        val efficiency = ItemHelper.getEnchantment(Enchantments.EFFICIENCY, level.registryAccess())

        for (player in gameHandle.participants) {
            val stack = unbreakable(ItemStack(Items.IRON_SHOVEL))
            stack.enchant(efficiency, 4)

            val inventory = player.inventory
            inventory.setItem(4, stack)
            PlayerInventoryAccess.setSelectedSlot(player, 4)
        }
    }

    private fun placeChest() {
        val box = MapUtil.readBox(map.requireProperty("chest-area"))
        val chestAreaList = ArrayList<BlockPos>()

        for (block in box) {
            val state = level.getBlockState(block)
            if (materials.contains(state)) {
                chestAreaList.add(block.immutable())
            }
        }

        val chestPos = chestAreaList[random.nextInt(chestAreaList.size)]
        level.setBlockAndUpdate(chestPos, Blocks.CHEST.defaultBlockState())
    }
}
